### PR TITLE
Add legend icon and legend in gmf-layertree

### PR DESCRIPTION
**Pull request by ger-benjamin**
Example : http://ger-benjamin.github.io/ngeo/master/examples/contribs/gmf/layertree.html
Relative to c2c issue: https://github.com/camptocamp/c2cgeoportal/issues/1707

### CODE REVIEW

File: contribs/gmf/src/directives/layertree.js [hunk 1]
```diff
@@ -268,6 +269,120 @@ gmf.LayertreeController.prototype.getNodeState = function(treeCtrl) {
...
+
+/**
+ * Get the WMTS legend URL for the given layer.
+ * @param {ol.layer.Tile} layer Tile layer as returned by the
+ * gmf layerHelper service.
+ * @return {?string} The legend URL or null.
+ * @private
+ */
+gmf.LayertreeController.prototype.getWMTSLegendUrl_ = function(layer) {
+  // FIXME case of multiple styles ?  case of multiple legendUrl ?
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/layertree.js]: I don't have any example of WMTS layers with multiple styles and, or, legend URL. So I don't know what I must do in this case...

**Comment by ger-benjamin** [comment]
Perhaps, we can move methods `gmf.LayertreeController.prototype.getWMTSLegendUrl_` and `gmf.LayertreeController.prototype.getWMSLegendUrl_` in another file ?

File: contribs/gmf/src/directives/partials/layertree.html [hunk 2]
```diff
@@ -1,10 +1,19 @@
 <span ng-if="::!layertreeCtrl.isRoot"
     ng-class="['depth-' + layertreeCtrl.depth, gmfLayertreeCtrl.getResolutionStyle(layertreeCtrl.node), gmfLayertreeCtrl.getNoSourceStyle(layertreeCtrl)]">
   <a href ng-click="::gmfLayertreeCtrl.toggleActive(layertreeCtrl)">
+    <span class="layerIcon">
```
Comment (fgravin) [inline contribs/gmf/src/directives/partials/layertree.html]: if there is no image ?

File: contribs/gmf/src/directives/partials/layertree.html [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
+      <img ng-src="{{gmfLayertreeCtrl.getLegendIconUrl(layertreeCtrl)}}"></img>
```
Comment (fgravin) [inline contribs/gmf/src/directives/partials/layertree.html]: single time binding ?

File: contribs/gmf/src/directives/partials/layertree.html (hunk 2 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/layertree.html]: I use this class to align the line with others (this `span` element has always a width of 20px).

File: contribs/gmf/src/directives/partials/layertree.html (hunk 3 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/layertree.html]: The icon change with the map scale no ?

File: contribs/gmf/src/directives/partials/layertree.html (hunk 2 above)
Comment (fgravin) [inline contribs/gmf/src/directives/partials/layertree.html]: You are right

File: contribs/gmf/src/directives/partials/layertree.html (hunk 3 above)
Comment (fgravin) [inline contribs/gmf/src/directives/partials/layertree.html]: You are right

Comment (fgravin) [inline contribs/gmf/src/directives/partials/layertree.html]: So it does a request per layer each time the resolution change.
It is a lot of requests.

**Comment by fgravin** [comment]
Looks good.
Just a comment on the number of requests made, but i don't think we can avoid that ...
Maybe we can avoid the request if the legend image is collapsed ?

**Comment by ger-benjamin** [comment]
@fgravin The good news is that I saw that in CGXP, we don't watch on current scale to set the little icon's layer image (what I call a legendIcon). We use a url without scale. So I can request my image with a one-time binding.

The bad news is that I don't know how to display legend only when its visible. Display it (or not) initially is easy. But I need to have a `ng-click` function to set the state of the node's image. And whatever I do, request an image as result of the `ng-click` function cause troubles with the collapse animation...

A `ng-if` (and no `ng-click`) on the `img` (or at a different place) doesn't work because... I need a `ng-click` to revalue the `ng-if`...

**Comment by fgravin** [comment]
In the geomapfish demo, when the legend panel is open, then it request legend image on scale change.

A ng-if (and no ng-click) on the img (or at a different place) doesn't work because... I need a ng-click to revalue the ng-if...



> A ng-if (and no ng-click) on the img (or at a different place) doesn't work because... I need a ng-click to revalue the ng-if...

Forget it.
The only way is to observe the `collapsed` class value of the `data-toggle`.
You have the $attrs for a reference on the html node. But you would need to add something in ngeo layer tree, but i don't see how to make something generic right of it.

**Comment by fgravin** [comment]
Be sure about the legend image request on scale change, see from spec or with @sbrunner.
When it's fixed, you can merge.

**Comment by sbrunner** [comment]
- The request on the layer icon (using a rule) don't depends on the scale,
- The request for the legend depends on the scale,

**Pull request merged**